.github/workflows/main.yml

```diff
@@ -1,25 +1,103 @@
-# This workflow will later be replaced with logic to "Generate Website Data"
-# The verify-gdrive.yml workflow file will be renamed to this one
-# We have to introduce this change in steps because GitHub gets confused until
-# we add the new workflow file to the master branch
 name: "Generate Website Data"
 on:
   workflow_dispatch:
+  push:
 jobs:
-  generate:
+  build:
     runs-on: ubuntu-latest
-    env:
-      REPO_OWNER: ${{ github.repository_owner}}
-      REPO_BRANCH: ${{ github.ref_name }}
-      SERVICE_ACCOUNT_KEY_JSON: ${{ secrets.SERVICE_ACCOUNT_KEY_JSON }}
-      GDRIVE_FOLDER: ${{ vars.GDRIVE_FOLDER }}
+    outputs:
+      devcontainer: ${{ steps.filter.outputs.devcontainer }}
+      noncontainer: ${{ steps.filter.outputs.noncontainer }}
     steps:
     - uses: actions/checkout@v3
-    - run: pip install -r gdrive_requirements.txt
-    - run: python test_pull_from_gdrive.py
-    - name: Archive pulled files
-      uses: actions/upload-artifact@v3
+    - name: Get changed files
+      id: changed-files
+      uses: tj-actions/changed-files@v40
+    - name: List all changed files
+      id: filter
+      run: |
+        devcontainer=false
+        noncontainer=true
+        for file in ${{ steps.changed-files.outputs.all_changed_files }}; do
+          echo "$file was changed"
+          if [[ ${{github.event_name}} = push ]]; then
+            if [[ $file = .devcontainer* ]]; then
+              devcontainer=true
+            elif [[ $file = *requirements.txt* ]]; then
+              devcontainer=true
+            elif [[ $file = Gemfile* ]]; then
+              devcontainer=true
+            fi
+          fi
+        done
+        
+        echo "devcontainer=$devcontainer" >> $GITHUB_OUTPUT
+        echo "noncontainer=$noncontainer" >> $GITHUB_OUTPUT
+    - name: Login to GitHub Container Registry
+      uses: docker/login-action@v3
       with:
-        name: redacted-netfile-files
-        path: .local/downloads
+        registry: ghcr.io
+        username: ${{github.actor}}
+        password: ${{secrets.GITHUB_TOKEN}}
+    - name: Build dev container
+      if: steps.filter.outputs.devcontainer == 'true'
+      run: |
+        docker build --no-cache --tag ghcr.io/caciviclab/disclosure-backend-static/${{github.ref_name}}:latest -f ./.devcontainer/Dockerfile .
+        docker push ghcr.io/caciviclab/disclosure-backend-static/${{github.ref_name}}:latest
```

Comment:
If I understand your first question, I think you're asking why we even use 'latest' instead of just using the branch name.  The reason is mainly convention.  Typically, there's a version number or string, but I don't want to generate a new version on every commit.  I just want to replace the latest.

"GitHub Packages usage is free for public packages."  I don't think there's a limit and they actually don't let you delete it if it's popular.  Also, we are creating an image per branch and not per commit.  It gets overwritten per branch.

Reply:
It's a small thing, but to clarify my first question, in the form
`ghcr.io/caciviclab/disclosure-backend-static/{ref_name}:latest`
`ghcr.io` is the registry, `latest` is the tag, and all of `caciviclab/disclosure-backend-static/{ref_name}` is the container repository. Appending the ref_name to the repo name creates a new repo for every ref_name, rather than tagging each image with the ref_name. it's a bit of an odd convention, but I don't think it really matters for us because I don't think we have a need to browse container images à la Dockerhub. In fact, I can't even find how to browse public images on GH Packages.